### PR TITLE
load(): validate publish_date, no independent publishers, no amz/bwb without ISBN

### PR DESCRIPTION
**Pull request by scottbarnes**
There's no issue associated with this, but it addresses some of the issues @seabelis raised in #open-librarians-g.

UNLESS the item is a promise item, for imports that go through load(), this PR:
- Stop imports prior to ~~1500 CE~~ 1400 CE from Amazon and BWB;
- Stop imports of AMZ/BWB source_records if the record has no ISBN.
- Stop import of books published in future years; and
- Stop imports of Independently Published books.

If an import is a promise item, based on the source record, these validations are *not* applied.

Additionally, these validations do *not* apply to books added via `/books/add`.

Feature.

### Technical

The `load()` validation has been largely consolidated in its own function, `validate_record()`, in part because they're related, in part because `load()` has quite a lot going on already.

#### Required fields
`normalize_import_record(rec)` checks `rec` for missing fields and raises an error. I did not include that check in `validate_record(rec)` because `validate_record(rec)` does *not* apply to promise items, but the required fields *do* apply to promise items.

I find this a bit confusing and I think it gestures towards the idea that we are starting to have enough rules such that we are getting near the place were our rules are a bit like, "A & B but not C, unless D, in which case, A & C", and that is a bit hard to follow with this flow.

That said, I am not sure if this PR is the place to address that, and instead I think it should help inform discussions about improving the import flow.

However, if it is desirable to partially address this now, perhaps validation could be of the form:
```python
    if not is_promise_item(rec):
        validate_non_promise_item(rec)
    validate_record_universal(rec)    
    normalize_import_record(rec)

```
Then the required fields validation could be moved from `normalize_import_record` to `validate_record_universal`. Whether this is a good road to go too far down I don't know, because the next logical extension is `validate_promise_item`. I am not sure how many more specialized validation functions there might be with this approach.

#### Callers handle error raising
I opted to go with having the validation functions return `bool`s and the like, and having the caller always raise any necessary errors. My rationale was the functions exist merely to perform checks, and it's the caller that decides what to do. That said, I know there are arguments both ways, and I'm happy to move the `raise` statements within the relevant validation functions if that approach is preferred.

### Testing

`publish_date` has a future year.
```
❯ curl -X POST \
     -H "Content-Type: application/json" \
     -H "Cookie: $OL_COOKIE" \
     -d '{
        "title": "Les noirs et les rouges",
        "source_records": ["ia:lesnoirsetlesrou0000garl"],
        "authors": [{"name": "Alberto Garlini"}],
        "publishers": ["Gallimard"],
        "isbn_13": ["9782072702211"],
        "publish_date": "3001-01",
        "number_of_pages": 5
     }' \
     http://localhost:8080/api/import

{"success": false, "error_code": "unhandled-exception", "error": "PublishedInFutureYear(3001)"}
```

`publish_date` prior to 1400 CE with a non-Amazon/BWB source:
```
curl -X POST \
     -H "Content-Type: application/json" \
     -H "Cookie: $OL_COOKIE" \
     -d '{
        "title": "Les noirs et les rouges",
        "source_records": ["ia:lesnoirsetlesrou0000garl"],
        "authors": [{"name": "Alberto Garlini"}],
        "publishers": ["Gallimard"],
        "isbn_13": ["9782072702211"],
        "publish_date": "January 1, 1399",
        "number_of_pages": 5
     }' \
     http://localhost:8080/api/import

{"success": true, "edition": {"key": "/books/OL43522513M", "status": "matched"}, "work": {"key": "/works/OL31822851W", "status": "matched"}}
```

`publish_date` prior to 1400 CE WITH an Amazon/BWB source:
```
curl -X POST \
     -H "Content-Type: application/json" \
     -H "Cookie: $OL_COOKIE" \
     -d '{
        "title": "Les noirs et les rouges",
        "source_records": ["amazon:lesnoirsetlesrou0000garl"],
        "authors": [{"name": "Alberto Garlini"}],
        "publishers": ["Gallimard"],
        "isbn_13": ["9782072702211"],
        "publish_date": "January 1, 1399",
        "number_of_pages": 5
     }' \
     http://localhost:8080/api/import

{"success": false, "error_code": "unhandled-exception", "error": "PublicationYearTooOld(1399)"}
```

Amazon import without an ISBN (BWB behavior is the same):
```
❯ curl -X POST \
     -H "Content-Type: application/json" \
     -H "Cookie: $OL_COOKIE" \
     -d '{
        "title": "Les noirs et les rouges",
        "source_records": ["amazon:lesnoirsetlesrou0000garl"],
        "authors": [{"name": "Alberto Garlini"}],
        "publishers": ["Gallimard"],
        "isbn_13": [],
        "publish_date": "January 1, 1500",
        "number_of_pages": 5
     }' \
     http://localhost:8080/api/import

{"success": false, "error_code": "unhandled-exception", "error": "SourceNeedsISBN()"}
```

But other import sources can import without an ISBN:
```
❯ curl -X POST \
     -H "Content-Type: application/json" \
     -H "Cookie: $OL_COOKIE" \
     -d '{
        "title": "Les noirs et les rouges",
        "source_records": ["ia:lesnoirsetlesrou0000garl"],
        "authors": [{"name": "Alberto Garlini"}],
        "publishers": ["Gallimard"],
        "isbn_13": [],
        "publish_date": "January 1, 2001",
        "number_of_pages": 5
     }' \
     http://localhost:8080/api/import
{"success": true, "edition": {"key": "/books/OL43522513M", "status": "matched"}, "work": {"key": "/works/OL31822851W", "status": "matched"}}
```

Amazon import with ISBN:
```
❯ curl -X POST \
     -H "Content-Type: application/json" \
     -H "Cookie: $OL_COOKIE" \
     -d '{
        "title": "Les noirs et les rouges",
        "source_records": ["amazon:lesnoirsetlesrou0000garl"],
        "authors": [{"name": "Alberto Garlini"}],
        "publishers": ["Gallimard"],
        "isbn_13": ["9782072702211"],
        "publish_date": "January 1, 1500",
        "number_of_pages": 5
     }' \
     http://localhost:8080/api/import

{"success": true, "edition": {"key": "/books/OL43522513M", "status": "modified"}, "work": {"key": "/works/OL31822851W", "status": "matched"}}
```

Independently Published:
```
❯ curl -X POST \
     -H "Content-Type: application/json" \
     -H "Cookie: $OL_COOKIE" \
     -d '{
        "title": "Les noirs et les rouges",
        "source_records": ["amazon:lesnoirsetlesrou0000garl"],
        "authors": [{"name": "Alberto Garlini"}],
        "publishers": ["Independently Published"],
        "isbn_13": ["9782072702211"],
        "publish_date": "January 1, 1500",
        "number_of_pages": 5
     }' \
     http://localhost:8080/api/import

{"success": false, "error_code": "unhandled-exception", "error": "IndependentlyPublished()"}
```

Promise items can bypass these checks (independently published in year 1001 without an ISBN). E.g.:
```
❮ curl -X POST \
     -H "Content-Type: application/json" \
     -H "Cookie: $OL_COOKIE" \
     -d '{
        "title": "Les noirs et les rouges",
        "source_records": ["promise:lesnoirsetlesrou0000garl"],
        "authors": [{"name": "Alberto Garlini"}],
        "publishers": ["Independently Published"],
        "isbn_13": [],
        "publish_date": "January 1, 1001",
        "number_of_pages": 5
     }' \
     http://localhost:8080/api/import
{"success": true, "edition": {"key": "/books/OL13M", "status": "matched"}, "work": {"key": "/works/OL4W", "status": "matched"}}
```

But promise items still need titles, as before:
```
❮ curl -X POST \
     -H "Content-Type: application/json" \
     -H "Cookie: $OL_COOKIE" \
     -d '{
        "source_records": ["promise:lesnoirsetlesrou0000garl"],
        "authors": [{"name": "Alberto Garlini"}],
        "publishers": ["Independently Published"],
        "isbn_13": [],
        "publish_date": "January 1, 1001",
        "number_of_pages": 5
     }' \
     http://localhost:8080/api/import
{"success": false, "error_code": "invalid-value", "error": "1 validation error for Book: title:   field required (type=value_error.missing)"}
```

### Stakeholders

@jimchamp

### CODE REVIEW

**Comment by scottbarnes** [comment]
> @scottbarnes can we change the 1500 CE date to 1400 CE? 1500 will stop the import of some of the earliest printed books https://en.wikipedia.org/wiki/Incunable
> 
> I'm sure archive.org has scans of books from this period, and many libraries have important editions from this period.

@hornc, I suspect I may have misunderstood @seabelis' [request](https://docs.google.com/document/d/1KbAK7qpDtlXA5SPyM8dnQI2QuBaNLzx3zpMPftGkJ5Y/edit#heading=h.19g1zvmr4i92). @seabelis, did you intend for the < 1500 CE import limitation to apply only to Amazon imports? Regardless, what do you feel about 1400?

File: openlibrary/catalog/add_book/tests/test_add_book.py [hunk 1]
```diff
@@ -1179,3 +1184,52 @@ def test_add_identifiers_to_edition(mock_site) -> None:
...
+    ],
+)
+def test_validate_record(name, rec, error, expected) -> None:
+    _ = name  # Name is just used to make the tests easier to understand.
+
+    if error:
+        with pytest.raises(error):
+            validate_record(rec)
+    else:
+        assert validate_record(rec) is expected  # type: ignore [func-returns-value]
```
Comment (hornc) [inline openlibrary/catalog/add_book/tests/test_add_book.py:1235]: `name` could be used here if the assertation isn't met in the assert failed message.

Comment (jimchamp) [reply openlibrary/catalog/add_book/tests/test_add_book.py:1235]: @hornc could you elaborate? I'm not even sure that passing `name` is useful here.

Comment (scottbarnes) [reply openlibrary/catalog/add_book/tests/test_add_book.py]: @hornc, I updated the PR to reflect this suggestion. E.g.:
```
>       assert sorted(get_missing_fields(rec=rec)) == sorted(expected), f"Assertion failed for test: {name}"                                                                                                                                                                                
E       AssertionError: Assertion failed for test: Catches multiple missing required fields                                                                                                                                                                                                 
E       assert ['source_records', 'title'] == ['flop', 'source_records']                                                                                                                                                                                                                    
E         At index 0 diff: 'source_records' != 'flop'                                                                                                                                                                                                                                       
E         Use -v to get more diff                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                            
openlibrary/tests/catalog/test_utils.py:338: AssertionError
```

**Review by jimchamp — CHANGES_REQUESTED** [review]
Thanks @scottbarnes! I've suggested some changes.  Once those are implemented and we resolve the earliest publish year question, I'm happy to merge this.

File: openlibrary/catalog/add_book/tests/test_add_book.py [hunk 2]
```diff
@@ -1179,3 +1184,50 @@ def test_add_identifiers_to_edition(mock_site) -> None:
...
+            None,
+        ),
+    ],
+)
+def test_validate_record(name, rec, error, expected) -> None:
+    if error:
+        with pytest.raises(error):
+            validate_record(rec)
+    else:
+        assert validate_record(rec) is expected, f"Assertion failed for test: {name}"  # type: ignore [func-returns-value]
```
[the remark this answers is not in the document]
Comment (scottbarnes) [reply openlibrary/catalog/add_book/tests/test_add_book.py]: Good point! I made this change.

File: openlibrary/catalog/utils/__init__.py [hunk 3]
```diff
@@ -284,3 +286,95 @@ def mk_norm(s: str) -> str:
...
+    likelihood is high that this is bad data. So we don't want to import these.
+    """
+    return publish_year > datetime.datetime.now().year
+
+
+def publication_year_too_old(publish_year: int) -> bool:
+    """
+    Returns True if publish_year is < 1,500 CE, and False otherwise.
+    """
+    return publish_year < 1500
```
[the remark this answers is not in the document]
Comment (scottbarnes) [reply openlibrary/catalog/utils/__init__.py]: I made this change, both here and in the error that references the earliest year allowed.

File: openlibrary/catalog/utils/__init__.py [hunk 4]
```diff
@@ -1,5 +1,7 @@
+import datetime
 import re
-from re import Match
+from re import compile, Match
+from typing import cast, Mapping
```
[the remark this answers is not in the document]
Comment (scottbarnes) [reply openlibrary/catalog/utils/__init__.py]: `Match` is needed, but `cast` and `Mapping` are not. How embarrassing. :(
I removed the unused imports.

**Comment by LeadSongDog** [comment]
@scottbarnes Can we also block January 1 (for any year)? These date parts are almost entirely artifacts.

**Comment by seabelis** [comment]
Probably just a mis-statement on @cdrini 's part, but classic work adaptations are not something we are specifically aiming to block. Classic work reprints by "independently published" and print-on-demand publishers are.

**Comment by LeadSongDog** [comment]
> @LeadSongDog The various study guides and notes should be independent editions. I would love to have a way to link them to the original work though.

Like this? https://openlibrary.org/books/OL45873022M/Lewis_Main_Street

**Comment by seabelis** [comment]
> > @LeadSongDog The various study guides and notes should be independent editions. I would love to have a way to link them to the original work though.
> 
> Like this? https://openlibrary.org/books/OL45873022M/Lewis_Main_Street

No. That just makes it an edition of the original work. That should be restored to the work it belonged to previously. What I meant was it would be good to indicate relationships between works, and their derivatives, criticism, etc. I currently make manual links in the description, but that's not perfect as that's not really a description, it's a relationship.

**Comment by LeadSongDog** [comment]
Ok, so I made it an independent work, with an url linking to the subject work.

File: openlibrary/catalog/add_book/tests/test_add_book.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (scottbarnes) [reply openlibrary/catalog/add_book/tests/test_add_book.py]: I have somewhat un-made this change, as I changed the test to remove the `expected` value, since nothing is expected currently. That said, if you look at the current form and think the `expected` value should return, I will happily resurrect it and the `==` comparison.

**Comment by scottbarnes** [comment]
I have updated the comments/description at the very top to reflect the current state of the PR, or at least what it _should_ be. I also specifically raise an issue about how to handle the required fields in a rec, as they are arguably validation rules, but also **do** apply to promise items, and none of the validation rules added in this PR apply to promise items.

**Comment by tfmorris** [comment]
> If an import is a promise item, based on the source record, these validations are not applied.

Why the exception? I've never seen a definition of what a "promise item" is, but in my experience they are a big, if not the biggest, source of bad data.

**Comment by scottbarnes** [comment]
@tfmorris, I am hoping someone will jump in correct me, but I *think* the reason for the exception is that there isn't currently a separate way to track promise items.

As for what promise items are, I believe they are items that have been 'promised' to Internet Archive that Internet Archive does not yet have, but will have, and that Internet Archive will someday scan for archival. They can be tracked [here](https://archive.org/details/protodonationitems?sort=-addeddate&tab=collection).

I think there was an issue with items being packed up and shipped for scanning, only for it to later emerge that Internet Archive already had the item scanned. Adding these promise items to Open Library stopped books coming from Open Library and going for scanning when they had in fact already been scanned.

However, I am glad you have focused on this exception, because it rather undercuts much of what this PR might achieve. Perhaps it will turn out (1) I am wholly wrong about my understanding in a way that changes this PR, or (2) that there is a way to address the needs around promise items that can be addressed here, or (3) perhaps it will lead to a future PR to handle promise item tracking in a better way.

File: openlibrary/catalog/utils/__init__.py [hunk 5]
```diff
@@ -319,3 +324,111 @@ def expand_record(rec: dict) -> dict[str, str | list[str]]:
...
+
+    >>> get_publication_year('1999-01')
+    1999
+    >>> get_publication_year('January 1, 1999')
+    1999
+    """
+    if publish_date is None:
+        return None
+
+    pattern = compile(r"\b\d{4}(?!\d)\b")
```
Comment (cdrini) [inline openlibrary/catalog/utils/__init__.py:342]:
```suggestion
    pattern = compile(r"\b\d{4}\b")
```

Comment (cdrini) [reply openlibrary/catalog/utils/__init__.py:342]: Maybe import `re.year` from openlibrary/catalog/utils/__init__.py

File: openlibrary/catalog/utils/__init__.py [hunk 6]
```diff
@@ -319,3 +324,111 @@ def expand_record(rec: dict) -> dict[str, str | list[str]]:
...
+
+def published_in_future_year(publish_year: int) -> bool:
+    """
+    Return True if a book is published in a future year as compared to the
+    current year.
+
+    Some import sources have publication dates in a future year, and the
+    likelihood is high that this is bad data. So we don't want to import these.
+    """
+    return publish_year > datetime.datetime.now().year
```
Comment (cdrini) [inline openlibrary/catalog/utils/__init__.py:356]: @seabelis I'm wondering if we might want to add a bit of wiggle room here. I find a lot of valid books are imported a little before their release date. Eg https://openlibrary.org/search?q=first_publish_year%3A2024&mode=everything

Currently it will block imports with publication year greater than eg 2023. Do you think we should allow for a +1 margin here?

**Review by cdrini — APPROVED** [review]
This lgtm ! From a policy perspective, I think this is in line with what we've all discussed so looks good!

File: openlibrary/catalog/utils/__init__.py (hunk 6 above)
Comment (LeadSongDog) [reply openlibrary/catalog/utils/__init__.py:357]: @cdrini To my way of looking at it we should not import something that isn’t yet published. Editions that were once available but are no longer extant are a more subtle call, but I’d avoid auto-importing them. We already have a huge pile of crap to clean up from bogus POD that were put out by “independently published” or the many others of that ilk.
